source/util/huffman_codec.h

```diff
@@ -0,0 +1,258 @@
+    queue_vector.reserve(hist.size());
+    std::priority_queue<Node*, std::vector<Node*>,
+        std::function<bool(const Node*, const Node*)>>
+	    queue(LeftIsBigger, std::move(queue_vector));
+
+    // Put all leaves in the queue.
+    for (const auto& pair : hist) {
+      Node* node = CreateNode();
+      node->val = pair.first;
+      node->weight = pair.second;
+      queue.push(node);
+    }
+
+    // Form the tree by combining two subtrees with the least weight,
+    // and pushing the root of the new tree in the queue.
+    while (true) {
+      // We push a node at the end of each iteration, so the queue is never
+      // supposed to be empty at this point, unless there are no leaves, but
+      // that case was already handled.
+      assert(!queue.empty());
+      Node* right = queue.top();
+      queue.pop();
+
+      // If the queue is empty at this point, then the last node contains is
+      // the root of the complete Huffman tree.
+      if (queue.empty()) {
+        root_ = right;
+        break;
+      }
+
+      Node* left = queue.top();
+      queue.pop();
+
+      // Combile left and right into a new tree and push it into the queue.
+      Node* parent = CreateNode();
+      parent->weight = right->weight + left->weight;
+      parent->left = left;
+      parent->right = right;
+      queue.push(parent);
+    }
+
+    // Traverse the tree and form encoding table.
+    CreateEncodingTable();
+  }
+
+  // Prints the Huffman tree in the following format:
+  // w------w------'x'
+  //        w------'y'
+  // Where w stands for the weight of the node.
+  void PrintTree(std::ostream& out) {
+    PrintTreeInternal(out, root_, 0);
+  }
+
+  void PrintTable(std::ostream& out, bool print_weights = true) {
+    std::queue<std::pair<Node*, std::string>> queue;
+    queue.emplace(root_, "");
+
+    while (!queue.empty()) {
+      const Node* node = queue.front().first;
+      const std::string code = queue.front().second;
+      queue.pop();
+      if (!node->right && !node->left) {
+        out << node->val;
+        if (print_weights)
+            out << " " << node->weight;
+        out << " " << code << std::endl;
+      } else {
+        if (node->left)
+          queue.emplace(node->left, code + "0");
+
+        if (node->right)
+          queue.emplace(node->right, code + "1");
+      }
+    }
+  }
+
+  const std::unordered_map<Val, std::pair<uint64_t, size_t>>&
+      GetEncodingTable() const {
+    return encoding_table_;
+  }
+
+  // Encodes |val| and stores the value in the lower |num_bits| of |bits|.
+  bool Encode(const Val& val, uint64_t* bits, size_t* num_bits) {
+    auto it = encoding_table_.find(val);
+    if (it == encoding_table_.end())
+      return false;
+    *bits = it->second.first;
+    *num_bits = it->second.second;
+    return true;
+  }
+
+  // Reads bits one-by-one using callback |read_bit| until a match is found.
+  // Matching value is stored in |val|. Returns false if the stream terminates
```

Comment:
Have to explain who the callback works.  There are two bools returned: one by return and the other by pointer param.  I shouldn't have to read the body of this method to know how to make a read_bit callback.